File: apps/customer_flutter/lib/measurement/decimal.dart
// Decimal-safe helpers for measurement quantities (max 3 decimal places).
// Uses integer milli-units (×1000) to avoid floating-point artefacts.

const int kMeasurementScale = 1000;

class MeasurementDecimalParse {
  const MeasurementDecimalParse._ok(this.milli, this.normalized)
      : ok = true,
        reason = null;
  const MeasurementDecimalParse._fail(this.reason)
      : ok = false,
        milli = 0,
        normalized = '';

  final bool ok;
  final int milli;
  final String normalized;
  final String? reason;
}

MeasurementDecimalParse parseMeasurementDecimalStrict(Object? value) {
  if (value == null || value == '') {
    return const MeasurementDecimalParse._fail('missing');
  }

  if (value is num) {
    if (!value.isFinite) {
      return const MeasurementDecimalParse._fail('not_finite');
    }
    final milli = (value * kMeasurementScale).round();
    if (((milli / kMeasurementScale) - value).abs() > 1e-9) {
      return const MeasurementDecimalParse._fail('too_many_decimals');
    }
    return MeasurementDecimalParse._ok(milli, milliToNormalizedString(milli));
  }

  final s = value.toString().trim();
  if (s.isEmpty) return const MeasurementDecimalParse._fail('empty');
  if (!RegExp(r'^-?\d+(\.\d+)?$').hasMatch(s)) {
    return const MeasurementDecimalParse._fail('invalid_format');
  }
  final frac = RegExp(r'^-?\d+\.(\d+)$').firstMatch(s);
  if (frac != null && frac.group(1)!.length > 3) {
    return const MeasurementDecimalParse._fail('too_many_decimals');
  }
  final n = num.tryParse(s);
  if (n == null || !n.isFinite) {
    return const MeasurementDecimalParse._fail('not_finite');
  }
  final milli = (n * kMeasurementScale).round();
  return MeasurementDecimalParse._ok(milli, milliToNormalizedString(milli));
}

String coerceMeasurementDecimalString(Object? value, [String fallback = '1']) {
  final parsed = parseMeasurementDecimalStrict(value);
  if (parsed.ok) return parsed.normalized;
  if (value != null) {
    final again = parseMeasurementDecimalStrict(value.toString());
    if (again.ok) return again.normalized;
  }
  return fallback;
}

String milliToNormalizedString(int milli) {
  final neg = milli < 0;
  final abs = milli.abs();
  final whole = abs ~/ kMeasurementScale;
  final frac = abs % kMeasurementScale;
  if (frac == 0) return '${neg ? '-' : ''}$whole';
  final fracStr = frac.toString().padLeft(3, '0').replaceAll(RegExp(r'0+$'), '');
  return '${neg ? '-' : ''}$whole.$fracStr';
}

bool isIntegerMilli(int milli) => milli % kMeasurementScale == 0;

int baseMilliToDisplayMilli(
  int baseMilli,
  String baseUnitCode,
  String displayUnitCode,
) {
  if (baseUnitCode == 'kg' && displayUnitCode == 'g') {
    return baseMilli * 1000;
  }
  if (baseUnitCode == 'l' && displayUnitCode == 'ml') {
    return baseMilli * 1000;
  }
  return baseMilli;
}

/// Add two normalized decimal strings via milli arithmetic.
String addQuantityDecimals(String a, String b) {
  final pa = parseMeasurementDecimalStrict(a);
  final pb = parseMeasurementDecimalStrict(b);
  if (!pa.ok || !pb.ok) return a;
  return milliToNormalizedString(pa.milli + pb.milli);
}

/// Subtract [step] from [value]; returns null if result would be < minimum.
String? subtractQuantityDecimal(String value, String step, String minimum) {
  final pv = parseMeasurementDecimalStrict(value);
  final ps = parseMeasurementDecimalStrict(step);
  final pm = parseMeasurementDecimalStrict(minimum);
  if (!pv.ok || !ps.ok || !pm.ok) return null;
  final next = pv.milli - ps.milli;
  if (next < pm.milli) return null;
  return milliToNormalizedString(next);
}

File: packages/core/src/measurement/decimal.ts
/**
 * Decimal-safe helpers for measurement quantities (max 3 decimal places).
 * Uses integer milli-units (×1000) to avoid floating-point artefacts.
 */

const SCALE = 1000;

export function parseMeasurementDecimalStrict(
  value: unknown
): { ok: true; milli: number; normalized: string } | { ok: false; reason: string } {
  if (value == null || value === '') return { ok: false, reason: 'missing' };

  let s: string;
  if (typeof value === 'number') {
    if (!Number.isFinite(value)) return { ok: false, reason: 'not_finite' };
    // Detect excessive precision via scaled integer check
    const milli = Math.round(value * SCALE);
    if (Math.abs(milli / SCALE - value) > 1e-9) return { ok: false, reason: 'too_many_decimals' };
    return { ok: true, milli, normalized: milliToNormalizedString(milli) };
  }

  s = String(value).trim();
  if (!s) return { ok: false, reason: 'empty' };
  if (!/^-?\d+(\.\d+)?$/.test(s)) return { ok: false, reason: 'invalid_format' };

  const fracMatch = s.match(/^-?\d+\.(\d+)$/);
  if (fracMatch && fracMatch[1].length > 3) return { ok: false, reason: 'too_many_decimals' };

  const n = Number(s);
  if (!Number.isFinite(n)) return { ok: false, reason: 'not_finite' };
  const milli = Math.round(n * SCALE);
  return { ok: true, milli, normalized: milliToNormalizedString(milli) };
}

/** Normalize Prisma Decimal / unknown to string without rejecting. Falls back to "1". */
export function coerceMeasurementDecimalString(value: unknown, fallback = '1'): string {
  const parsed = parseMeasurementDecimalStrict(value);
  if (parsed.ok) return parsed.normalized;
  if (value != null && typeof (value as { toString?: () => string }).toString === 'function') {
    const again = parseMeasurementDecimalStrict(String(value));
    if (again.ok) return again.normalized;
  }
  return fallback;
}

export function milliToNormalizedString(milli: number): string {
  const neg = milli < 0;
  const abs = Math.abs(milli);
  const whole = Math.floor(abs / SCALE);
  const frac = abs % SCALE;
  if (frac === 0) return `${neg ? '-' : ''}${whole}`;
  const fracStr = String(frac).padStart(3, '0').replace(/0+$/, '');
  return `${neg ? '-' : ''}${whole}.${fracStr}`;
}

export function milliToNumber(milli: number): number {
  return milli / SCALE;
}

export function isIntegerMilli(milli: number): boolean {
  return milli % SCALE === 0;
}

/** Convert base-unit milli to display milli (g/ml = ×1000). */
export function baseMilliToDisplayMilli(
  baseMilli: number,
  baseUnitCode: string,
  displayUnitCode: string
): number {
  if (baseUnitCode === 'kg' && displayUnitCode === 'g') return baseMilli * 1000;
  if (baseUnitCode === 'l' && displayUnitCode === 'ml') return baseMilli * 1000;
  return baseMilli;
}
